### PR TITLE
[GEOT-6264] JDBC FeatureWriter-implementation returns null values in getAttributes()

### PR DESCRIPTION
**Pull request by bestrauss**
@aaime

Here my next try. I switched off code-formatting etc. Changes now reduce to two lines in one file (plus two new unit-test files). Following contribution guidelines, "Talk first policy" does not apply to "easy bugfix with test".

See Jira GEOT-6264 for the Version of JDBCFeatureWriterOnlineTest.java which demonstrates the bug. The JDBC FeatureWriter implementation returns features whose attributes are fine in case obtained via getAttribute(int) but which are null in case obtained via getAttributes().

Please note also: I added .gradle to .gitignore.

### CODE REVIEW

**Comment by bestrauss** [comment]
The Travis CI build failed. Here's the: [log](https://api.travis-ci.org/v3/job/505217031/log.txt). Reason is: 

> test failures on project gt-jdbc-h2
> JDBCFeatureSourceOnlineTest.testRendererBehaviour:382 Â» NullPointer

The featureSource is null. The dataStore does not return a featureSource. I don't see how this could be related to my changes, since a FeatureWriter or FeatureReader has not been instantiated. I would have to learn about H2 Databases and install one to continue.

File: modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureWriterOnlineTest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * JDBC {@link FeatureWriter}-implementation tampers with input-data.
+ *
+ * @author Burkhard Strauss
+ * @see GEOT-6264
+ */
+public abstract class JDBCFeatureWriterOnlineTest extends JDBCTestSupport {
+
+    public void testNext() throws Exception {
+
+        try (FeatureWriter writer =
```
Comment (aaime) [inline modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureWriterOnlineTest.java]: The description states "The test shows that attributes of features returned by JDBC-FeatureWriters are ok if retrieved via {{feature.getAttribute(index)}} but are {{null}} if retrieved via {{feature.getAttributes()}}."
But I don't see any call to getAttributes() here?

**Comment by aaime** [comment]
H2 is nothing you have to install, it's an embedded database, tests in H2 run by default precisely because there is no need to configure any connection. Travis does not check any other DB because we don't have the databases to use for tests (and the build would be too slow, too).

I've restarted the failing build just in case, but if it keeps failing, it may be a side effect of your changes. You can run the test locally with no issue.

File: modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureWriterOnlineTest.java (hunk 1 above)
Comment (bestrauss) [reply modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCFeatureWriterOnlineTest.java]: Your're right my fault. The original test was to demonstrate the problem. The modified test must call getAttributes() and ensure they're not null. I will correct that.

**Comment by bestrauss** [comment]
I updated the test to call getAttributes().

H2 test runs successfully on my local machine. I had thought this might be for the same reason, MySQL Tests are successful without a database configured.

**Comment by aaime** [comment]
MySQL tests are skipped without a database :-)

Il mar 12 mar 2019, 20:31 Burkhard Strauss <notifications@github.com> ha
scritto:

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCFeatureReader.java [hunk 2]
```diff
@@ -647,6 +647,9 @@ public void init() throws SQLException, IOException {
 
             // TODO: factory fid prefixing out
             init(featureType.getTypeName() + "." + dataStore.encodeFID(key, rs, offset));
+
+            for (int k = 0; k < values.length; k++) // ensure initialized values GEOT-6264
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCFeatureReader.java]: The review I made last week, but found out this morning I forgot to send it over... sorry!

[the remark this answers is not in the document]
Comment (bestrauss) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCFeatureReader.java]: Yes, it does make getAttributes() usable again.

[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCFeatureReader.java]: Ok, let's merge then, yes?

[the remark this answers is not in the document]
Comment (bestrauss) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCFeatureReader.java]: Go ahead!

**Comment by aaime** [comment]
Ok, it's in. Want to make backport PRs too? If so, the branches currently alive are 21.x and 20.x.

**Pull request merged**